src/mscorlib/src/System/Threading/ExecutionContext.cs

```diff
@@ -191,18 +191,27 @@ internal static void SetLocalValue(IAsyncLocal local, object newValue, bool need
             if (previousValue == newValue)
                 return;
 
-            current = new ExecutionContext(current);
-            current.m_localValues[local] = newValue;
-
-            t_currentMaybeNull = current;
+            Dictionary<IAsyncLocal, object> newValues = new Dictionary<IAsyncLocal, object>(current.m_localValues);
```

Comment:
Dictionary's ctor takes an IDictionary, so enumerating this is going to allocate an enumerator object.  Justin's proposed change at https://github.com/dotnet/coreclr/pull/1620 would help with that.  And if there's a good chance m_localValues is empty at this point, we could also just do a count check and use the Dictionary ctor overload that doesn't take an input.